### DISCUSSION TITLE
Add tool/equipment-list for a recipe

### DISCUSSION THREAD

**Discussion by Knalltuete5000** [TandoorRecipes/recipes#2062] [Ideas]
Since some recipes require special tools it would be helpfull to provide a list of tools for each recipe.
This feature was already requested but never implemented (#301) 

First of all the tool list will be optional for recipes.
This feature will include at least two new models (tool and toolset) as well as a new foreign key for a recipe.

The tool model could contain following information:

- id (autogenerated, required)
- name (chars, required)
- description (chars, optional)

The toolset model could contain following information:

- id (autogenerated, required)
- amount (integer, optional)
- unitid (foreign-key -> unit, optional)
- toolid (foreign-key -> tool, required)
- note (chars, optional)
- no_amount (default false, does not show amount and the unit)
- no_unit (default false, does not show the unit but the amount)

As you might have noticed the models are similar to the food and ingredient models and should contain the missing information (e.g. space).

Like described in the toolset model I would like to introduce two options for the toolset: 

1. to only show the amount and the name of the tool. 
2. to only show the name of the tool.

I hope I have described the feature/addtion as well as possible and  would like to discuss this feature/addition with you but also would like your feedback on what is missing or should be removed from the feature.

I am willing to implement this feature if it is approved and welcomed.

And I would like to start the discussion with the following questions:

- can a tool be included to the shopping list? Or should this feature be added in a future release?
- Should there be any additional options for the toolset or the tool itself?
- Onto wording:
  - should it be called tool. Another option would be equipment, ...
  - same question for the toolset

**Comment by vabene1111** [comment]
Thanks for the well written proposal, that's exactly what I think gets a good feature/PR started. 

Before answering your questions I want to prepend one: Do we really need two models/that much relational data
Reasoning: While we as tech savvy people love our structured data and don't mind entering some additional information the majority of people using tandoor (e.g. my mum, my girlfriend, ...) do not love entering data into forms as much as I do. Tandoor has a huge problem with complexity to a degree where even tech savvy people I have showed the application to were not able to use it because it was not understandable enough. 

Long story short: When I have seen "tools" or "equipment" being implemented somewhere it has always been just a comma seperated list/multiselect. I do think that adding additional information is nice but I think i would limit it to an amount field so that we can save ourselfs the unit. Also if we only have amount we can automate the logic for `no_amount` since its either 1 (dont show anythign) or more (show the amount). 

When it comes to the note I am unsure, we could leave that in, but the `toolset` needs to be "invisible" to the user, basically just the relation, not something he has to create manually. 


regarding your questions
- shopping: interesting and nice idea, maybe we should add a user based tool relation to check if someone has a tool or not. we could then list the tools in the "add to shopping" dialog and ask the user if they want to add them or if they have them. I would implement this in another PR after having tools around for a few weeks and testing them 
- additional options, all optional: 
    - empty weight of the tool (for a bowl for example, this allows me to weigh it and now how much is in it if i forgot to zero the scale before filling it)
    - capacity (amount, unit)
    - location (were its stored in the kitchen)
    - quantity (how many of a tool you have)
    - id (probably wont be something many people have but if you want to track some kind of serial number, id, ...)
    - Now that i am writing this its becoming more and more apparent that this is not a base attribute but should be a user relation (or better a household relation but that does not exist yet, maybe this could be combined with what I wrote for shopping above)

Names: 
In German it would be "Utensilien" which translates to "Utensils", is that applicable to English kitchen wording? The toolset should be completely transparent to the user so the name does not matter that much, if the base was `Utensil` we could have the relations `UserUtensil` and `RecipeUtensil` (dont really like the sound of that, maybe someone has a better idea).

I dont really know how to make the user tool relation easy to use and understand for the user but i think it would be a very strong thing but we could also leave it out for now until I have implemented groups/houshoulds and add it afterwards so we dont have to write complex migrations. 

Looking forward to feedback

**Reply by Knalltuete5000** [reply]
Thanks for the response.

Great suggestion to remove the unit from the tool (I keep calling them tools for now, let's see what we end up with 😃) because in most cases you do not need them. But let's have a look on other input if there is any case where you need a unit but I don't see any at the moment.

For the shopping list I would prefer to make a second PR like you said or even have a look if it is request by the community and add it later.

For the additional options is it possible to create a list of optional options or even create a short description of the models in this discussion to keep the information centralized on what we agreed, so everybody can see them? And also it would be easier to create an issue/pr with a list of todo's.

**Reply by vabene1111** [reply]
yes, lets wait a few days to see if additional feedback comes in and if not we can create a model list. In the meanwhile I will think about how to make the user based tool relations easy to understand/transparent to the user so that he does not have to create multiple relations manually

**Reply by vabene1111** [reply]
ok it seems like no additional feedback is coming in. 

I am still not entirely sure how I would like to handle this, I am pretty sure that I want to add groups rather soon but I dont want to block this feature until then. 

I think for now we should add it without all that extra functionality, just a list of tools needed, adding the ability to decide what a user has can be added at a later stage once groups are done.

**Comment by MattSidney** [comment]
I don't think I have much to add other than my support for this idea. 
I have long stored equipment information in Notes or other parts of (various) recipe mangers. 

The idea of a dedicated section where equipment could be listed would be great. I do a lot of cooking at friends or families or on holiday so this would also simplify packing. 

Ideally it would be nice to select a number of recipies in order to generate a complete equipment list for packing/organisation purposes.

**Reply by vabene1111** [reply]
Definitely planned, really like the idea but have not yet gotten around to it

**Reply by Knalltuete5000** [reply]
Just to let you know: I have started to work on this feature [here](https://github.com/Knalltuete5000/tandoorecipes/tree/feature/equipment).
If you have the time and only if you can review the code. Else there will be a PR soon ™️  😅
